### PR TITLE
feat: use keccak builtin

### LINKED ISSUE #966: feat: use get_block_hash from cairo1
### Describe the Feature Request

The `get_block_hash` syscall is not available in Cairo0.

### Describe Preferred Solution

Write a Cairo1 contract to support syscall and other missing things in Cairo 0, to be used by Kakarot when required.
See https://book.cairo-lang.org/appendix-07-system-calls.html?highlight=syscall#get_block_hash

### PR DESCRIPTION
**Pull request by enitrat**
Time spent on this PR: 0.6d

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #966 

## What is the new behavior?

-
-
-

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1053)

### CODE REVIEW

File: tests/src/kakarot/instructions/test_environmental_information.py [hunk 1]
```diff
@@ -113,6 +116,25 @@ def test_gasprice(self, cairo_run):
...
+                if (bytes8_little_endian and last_word_bytes_used != 0)
+                else 0
+            )
+
+            full_words = []
+            if len(bytecode) % 8 == 0:
+                full_words = bytes8_little_endian
+            else:
+                full_words = bytes8_little_endian[:-1]
+            return len(full_words), full_words, last_word, last_word_bytes_used
```
Comment (ClementWalter) [inline tests/src/kakarot/instructions/test_environmental_information.py:136]: is this chat gpt generated? ;) could be simplified

```
if len(bytecode) % 8:
    last_word = bytes8_little_endian[-1]
    full_words = bytes8_little_endian[:-1]
else:
    last_word = 0
    full_words = bytes8_little_endian
```

File: tests/src/kakarot/instructions/test_environmental_information.py [hunk 2]
```diff
@@ -128,13 +150,38 @@ class TestExtCodeHash:
...
+            ), SyscallHandler.patch(
+                "ICairo1Helpers.library_call_keccak",
+                lambda class_hash, data: int_to_uint256(bytecode_hash),
             ):
                 output = cairo_run("test__exec_extcodehash", address=address)
-            assert output == (
-                hex(bytecode_hash) if address == EXISTING_ACCOUNT else "0x0"
-            )
+
+                if address == EXISTING_ACCOUNT:
```
Comment (ClementWalter) [inline tests/src/kakarot/instructions/test_environmental_information.py:168]: you can deindent here (leave the context manager)

File: tests/src/utils/test_bytes.py [hunk 3]
```diff
@@ -59,14 +59,63 @@ def test_should_return_bytes(self, cairo_run, n):
...
             ]
-            output = cairo_run(
-                "test__bytes_to_bytes8_little_endian", bytes=list(range(bytes_len))
-            )
-            assert bytes8_little_endian == output
+            # Case where the last word is not full
+            if bytes8_little_endian and not len(bytes_array) % 8 == 0:
+                last_expected_word = bytes8_little_endian[-1]
+                last_expected_word_bytes_used = (
+                    divmod(last_expected_word.bit_length(), 8)[0] + 1
```
Comment (ClementWalter) [inline tests/src/utils/test_bytes.py:74]: can use `last_expected_word.bit_length() // 8`

File: tests/src/kakarot/instructions/test_environmental_information.py (hunk 1 above)
Comment (enitrat) [reply tests/src/kakarot/instructions/test_environmental_information.py:136]: I wrote this 😢

Comment (ClementWalter) [reply tests/src/kakarot/instructions/test_environmental_information.py]: it's just that you don't make it visible that there is here one single condition based on the len of the bytecode, leading to either full words or remainders

Comment (enitrat) [reply tests/src/kakarot/instructions/test_environmental_information.py]:
```py
            last_word_bytes_used = len(bytecode) % 8
            last_word = (
                bytes8_little_endian[-1]
                if bytes8_little_endian and last_word_bytes_used
                else 0
            )
            full_words = (
                bytes8_little_endian[:-1]
                if last_word_bytes_used
                else bytes8_little_endian
            )
 ```
 
 How about this one ?

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**